base/stat.jl

```diff
@@ -104,182 +108,97 @@ The fields of the structure are:
-
-julia> isfile(f)
-true
-
-julia> close(f); rm("test_file.txt")
-```
-"""
-isfile(st::StatStruct) = filemode(st) & 0xf000 == 0x8000
-
-"""
-    islink(path) -> Bool
-
-Returns `true` if `path` is a symbolic link, `false` otherwise.
-"""
-islink(st::StatStruct) = filemode(st) & 0xf000 == 0xa000
-
-"""
-    issocket(path) -> Bool
-
-Returns `true` if `path` is a socket, `false` otherwise.
-"""
-issocket(st::StatStruct) = filemode(st) & 0xf000 == 0xc000
-
-# mode permission predicates
-
-"""
-    issetuid(path) -> Bool
-
-Returns `true` if `path` has the setuid flag set, `false` otherwise.
-"""
-issetuid(st::StatStruct) = (filemode(st) & 0o4000) > 0
-
-"""
-    issetgid(path) -> Bool
+function stat(path..., link = false)
+    full_path = joinpath(path...)
+    if link
+        stat(full_path)
+    else
+         lstat(full_path)
+    end
+end
 
-Returns `true` if `path` has the setgid flag set, `false` otherwise.
 """
-issetgid(st::StatStruct) = (filemode(st) & 0o2000) > 0
+    filetype(path; link = false) -> Symbol
+
+Returns the file type of `path`, with the following possibilities:
+
+- `:invalid`: not a valid filesystem path.-
+- `:FIFO`: a FIFO
+- `:chardev`: a character device
+- `:dir`: a directory
+- `:blockdev`: a block device
+- `:file`: a regular file
+- `:link`: a symbolic link
+- `:socket`: a socket
+
+If `link` is true, get the type of a link itself, with the same restrictions as
+[`stat`](@ref).
+"""
+function filetype(st::StatStruct; link = false)
+    mode = st.mode & 0xf000
+    if mode == 0x0000
+        :invalid
+    elseif mode == 0x1000
+        :FIFO
+    elseif mode == 0x2000
+        :chardev
+    elseif mode == 0x4000
+        :dir
+    elseif mode == 0x6000
+        :blockdev
+    elseif mode == 0x8000
+        :file
+    elseif mode == 0xa000
+        :link
+    elseif mode == 0xc000
+        :socket
+    else
+        error("Unknown path type")
+    end
+end
 
 """
-    issticky(path) -> Bool
+    fileflags(path)
 
-Returns `true` if `path` has the sticky bit set, `false` otherwise.
-"""
-issticky(st::StatStruct) = (filemode(st) & 0o1000) > 0
+Returns a named tuple of the flags of a `path`, if the `path`:
 
+- `setuid`: has the setuid flag set
+- `setgid`: has the setgid flag set
+- `sticky`: has the sticky bit set
 """
-    uperm(file)
-
-Gets the permissions of the owner of the file as a bitfield of
-
-| Value | Description        |
-|:------|:-------------------|
-| 01    | Execute Permission |
-| 02    | Write Permission   |
-| 04    | Read Permission    |
+fileflags(path) = (
+    setuid = (filemode(st) & 0o4000) > 0,
+    setgid = (filemode(st) & 0o2000) > 0,
+    sticky = (filemode(st) & 0o1000) > 0
+)
 
-For allowed arguments, see [`stat`](@ref).
-"""
-uperm(st::StatStruct) = UInt8((filemode(st) >> 6) & 0x7)
+translate_permission(bit) =
+    if bit == 1
+        :execute
+    elseif bit == 2
+        :write
+    elseif bit == 4
+        :read
+    else
+        error("Unknown permissions type")
+    end
```

Comment:
This doesn't handle multiple permission bits being set.

Reply:
Oh, ok, I must have misunderstood the code. How would I fix it?

Reply:
I thought I would be ok with `permissions` returning a named tuple for all 3.

Reply:
That part is ok, but each of user/group/other might have any combination of read/write/execute.